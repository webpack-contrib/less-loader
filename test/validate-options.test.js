import { getCompiler, compile } from './helpers/index';

describe('validate options', () => {
  const tests = {
    lessOptions: {
      success: [
        { strictMath: true },
        () => ({
          strictMath: true,
        }),
      ],
      failure: [1, true, false, 'test', []],
    },
    prependData: {
      success: ['@background: coral;', () => '@background: coral;'],
      failure: [1, true, false, /test/, [], {}],
    },
    appendData: {
      success: ['@background: coral;', () => '@background: coral;'],
      failure: [1, true, false, /test/, [], {}],
    },
    sourceMap: {
      success: [true, false],
      failure: ['string'],
    },
    implementation: {
      // eslint-disable-next-line global-require
      success: [require('less')],
      failure: [true, false, 'string'],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === 'object' && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === 'success' ? 'successfully validate' : 'throw an error on'
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      const compiler = getCompiler('./basic.less', {
        [key]: value,
      });
      let stats;

      try {
        stats = await compile(compiler);
      } finally {
        if (type === 'success') {
          expect(stats.hasErrors()).toBe(false);
        } else if (type === 'failure') {
          const {
            compilation: { errors },
          } = stats;

          expect(errors).toHaveLength(1);
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
