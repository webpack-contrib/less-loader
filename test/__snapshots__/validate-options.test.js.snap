// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "appendData" option with "/test/" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.appendData should be one of these:
   string | function
   -> Add \`Less\` code after the actual entry file (https://github.com/webpack-contrib/less-loader#postponeddata).
   Details:
    * options.appendData should be a string.
    * options.appendData should be an instance of function."
`;

exports[`validate options should throw an error on the "appendData" option with "[]" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.appendData should be one of these:
   string | function
   -> Add \`Less\` code after the actual entry file (https://github.com/webpack-contrib/less-loader#postponeddata).
   Details:
    * options.appendData should be a string.
    * options.appendData should be an instance of function."
`;

exports[`validate options should throw an error on the "appendData" option with "{}" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.appendData should be one of these:
   string | function
   -> Add \`Less\` code after the actual entry file (https://github.com/webpack-contrib/less-loader#postponeddata).
   Details:
    * options.appendData should be a string.
    * options.appendData should be an instance of function."
`;

exports[`validate options should throw an error on the "appendData" option with "1" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.appendData should be one of these:
   string | function
   -> Add \`Less\` code after the actual entry file (https://github.com/webpack-contrib/less-loader#postponeddata).
   Details:
    * options.appendData should be a string.
    * options.appendData should be an instance of function."
`;

exports[`validate options should throw an error on the "appendData" option with "false" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.appendData should be one of these:
   string | function
   -> Add \`Less\` code after the actual entry file (https://github.com/webpack-contrib/less-loader#postponeddata).
   Details:
    * options.appendData should be a string.
    * options.appendData should be an instance of function."
`;

exports[`validate options should throw an error on the "appendData" option with "true" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.appendData should be one of these:
   string | function
   -> Add \`Less\` code after the actual entry file (https://github.com/webpack-contrib/less-loader#postponeddata).
   Details:
    * options.appendData should be a string.
    * options.appendData should be an instance of function."
`;

exports[`validate options should throw an error on the "implementation" option with "false" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.implementation should be an object:
   object { … }
   -> The implementation of the \`Less\` to be used (https://github.com/webpack-contrib/less-loader#implementation)."
`;

exports[`validate options should throw an error on the "implementation" option with "string" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.implementation should be an object:
   object { … }
   -> The implementation of the \`Less\` to be used (https://github.com/webpack-contrib/less-loader#implementation)."
`;

exports[`validate options should throw an error on the "implementation" option with "true" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.implementation should be an object:
   object { … }
   -> The implementation of the \`Less\` to be used (https://github.com/webpack-contrib/less-loader#implementation)."
`;

exports[`validate options should throw an error on the "lessOptions" option with "[]" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.lessOptions should be one of these:
   object { … } | function
   -> Options to pass through to \`Less\` (https://github.com/webpack-contrib/less-loader#lessoptions).
   Details:
    * options.lessOptions should be an object:
      object { … }
    * options.lessOptions should be an instance of function."
`;

exports[`validate options should throw an error on the "lessOptions" option with "1" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.lessOptions should be one of these:
   object { … } | function
   -> Options to pass through to \`Less\` (https://github.com/webpack-contrib/less-loader#lessoptions).
   Details:
    * options.lessOptions should be an object:
      object { … }
    * options.lessOptions should be an instance of function."
`;

exports[`validate options should throw an error on the "lessOptions" option with "false" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.lessOptions should be one of these:
   object { … } | function
   -> Options to pass through to \`Less\` (https://github.com/webpack-contrib/less-loader#lessoptions).
   Details:
    * options.lessOptions should be an object:
      object { … }
    * options.lessOptions should be an instance of function."
`;

exports[`validate options should throw an error on the "lessOptions" option with "test" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.lessOptions should be one of these:
   object { … } | function
   -> Options to pass through to \`Less\` (https://github.com/webpack-contrib/less-loader#lessoptions).
   Details:
    * options.lessOptions should be an object:
      object { … }
    * options.lessOptions should be an instance of function."
`;

exports[`validate options should throw an error on the "lessOptions" option with "true" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.lessOptions should be one of these:
   object { … } | function
   -> Options to pass through to \`Less\` (https://github.com/webpack-contrib/less-loader#lessoptions).
   Details:
    * options.lessOptions should be an object:
      object { … }
    * options.lessOptions should be an instance of function."
`;

exports[`validate options should throw an error on the "prependData" option with "/test/" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.prependData should be one of these:
   string | function
   -> Prepends \`Less\` code before the actual entry file (https://github.com/webpack-contrib/less-loader#prependdata).
   Details:
    * options.prependData should be a string.
    * options.prependData should be an instance of function."
`;

exports[`validate options should throw an error on the "prependData" option with "[]" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.prependData should be one of these:
   string | function
   -> Prepends \`Less\` code before the actual entry file (https://github.com/webpack-contrib/less-loader#prependdata).
   Details:
    * options.prependData should be a string.
    * options.prependData should be an instance of function."
`;

exports[`validate options should throw an error on the "prependData" option with "{}" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.prependData should be one of these:
   string | function
   -> Prepends \`Less\` code before the actual entry file (https://github.com/webpack-contrib/less-loader#prependdata).
   Details:
    * options.prependData should be a string.
    * options.prependData should be an instance of function."
`;

exports[`validate options should throw an error on the "prependData" option with "1" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.prependData should be one of these:
   string | function
   -> Prepends \`Less\` code before the actual entry file (https://github.com/webpack-contrib/less-loader#prependdata).
   Details:
    * options.prependData should be a string.
    * options.prependData should be an instance of function."
`;

exports[`validate options should throw an error on the "prependData" option with "false" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.prependData should be one of these:
   string | function
   -> Prepends \`Less\` code before the actual entry file (https://github.com/webpack-contrib/less-loader#prependdata).
   Details:
    * options.prependData should be a string.
    * options.prependData should be an instance of function."
`;

exports[`validate options should throw an error on the "prependData" option with "true" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.prependData should be one of these:
   string | function
   -> Prepends \`Less\` code before the actual entry file (https://github.com/webpack-contrib/less-loader#prependdata).
   Details:
    * options.prependData should be a string.
    * options.prependData should be an instance of function."
`;

exports[`validate options should throw an error on the "sourceMap" option with "string" value 1`] = `
"Invalid options object. Less Loader has been initialized using an options object that does not match the API schema.
 - options.sourceMap should be a boolean.
   -> Enables/Disables generation of source maps (https://github.com/webpack-contrib/less-loader#sourcemap)."
`;
